src/shell/shell_cmds.cpp

```diff
@@ -401,28 +402,38 @@ void DOS_Shell::CMD_RMDIR(char * args) {
 	}
 }
 
-static void FormatNumber(Bit32u num,char * buf) {
-	Bit32u numm,numk,numb,numg;
-	numb=num % 1000;
-	num/=1000;
-	numk=num % 1000;
-	num/=1000;
-	numm=num % 1000;
-	num/=1000;
-	numg=num;
-	if (numg) {
-		sprintf(buf,"%d,%03d,%03d,%03d",numg,numm,numk,numb);
-		return;
-	};
-	if (numm) {
-		sprintf(buf,"%d,%03d,%03d",numm,numk,numb);
-		return;
-	};
-	if (numk) {
-		sprintf(buf,"%d,%03d",numk,numb);
-		return;
-	};
-	sprintf(buf,"%d",numb);
+static std::string format_number(size_t num)
+{
+	constexpr size_t petabyte_si = 1'000'000'000'000'000;
+	assert(num <= petabyte_si);
+	const auto b = static_cast<unsigned>(num % 1000);
+	num /= 1000;
+	const auto kb = static_cast<unsigned>(num % 1000);
+	num /= 1000;
+	const auto mb = static_cast<unsigned>(num % 1000);
+	num /= 1000;
+	const auto gb = static_cast<unsigned>(num % 1000);
+	num /= 1000;
+	const auto tb = static_cast<unsigned>(num);
+	char buf[22];
+	if (tb) {
+		safe_sprintf(buf, "%u,%03u,%03u,%03u,%03u", tb, gb, mb, kb, b);
+		return buf;
+	}
+	if (gb) {
+		safe_sprintf(buf, "%u,%03u,%03u,%03u", gb, mb, kb, b);
+		return buf;
+	}
+	if (mb) {
+		safe_sprintf(buf, "%u,%03u,%03u", mb, kb, b);
+		return buf;
+	}
+	if (kb) {
+		safe_sprintf(buf, "%u,%03u", kb, b);
+		return buf;
```

Comment:
http://wordpress.mrreid.org/2014/05/27/stop-putting-commas-in-your-numbers/

Suggest we start migrating to using spaces (when we make improvement like this) instead of commas as its the most universally understood metric notation.  Yes, plenty of coutrnies still incorrectly use a comma or period of out habit, and some teachers still incorrectly teach it.  But we don't have to carry on those flawed traditions.

Reply:
I will try how it looks, but we don't have thin-space character in DOS :(

I have no strong feelings about this, but I would prefer to keep the original formatting (except I suspect it shouldn't be a comma, but rather a locale-specific character - but I don't know for sure).

However for our `ls` command I have no such reservations (when I'll add `ls -l` support), here's how GNU ls behaves:

- by default, GNU `ls` avoids separators altogether (I find it quickly turns unreadable)

      $ ls -l | grep "README$"
      -rw-r--r--  1 dreamer_ dreamer_ 67498 Dec  8 10:40 README

- when asked for separator, it uses locale-appropriate one (I use en-GB locale for language, therefore it's `,` on my machine):

      $ BLOCK_SIZE="'1" ls -l | grep "README$"
      -rw-r--r--  1 dreamer_ dreamer_ 67,498 Dec  8 10:40 README

- When asked for "human readable" units, it uses `G,M,K` suffixes which mean Gibibyte,Mebibyte,Kibibyte (powers of 2) - (bad GNU!), and `.` as decimel separator (on my locale):

      $ ls -l -h | grep "README$"
      -rw-r--r--  1 dreamer_ dreamer_   66K Dec  8 10:40 README

- For SI units (power of 10), user needs to ask explicitly or change the defaults (it uses `G,M,k` suffixes now)

      $ ls -l -si
      $ BLOCK_SIZE="si" ls -l

Reply:
> except I suspect it shouldn't be a comma, but rather a locale-specific character - but I don't know for sure

Hmm.. right. I guess that would involve having a non-financial numerical dellimeter defined in all our locales (along with date, money symbol, etc).

I'm now thinking this is out of scope for the intent of this PR.. but a worthy change if/when someone wants to tackle locale formatting in general. Probably a small can of worms!

Reply:
Exactly. For what it's worth, I tested version of `dir` with spaces for file sizes - they are visually "merging" with dates, so another modification would be needed :( I will merge the change without touching it.

Reply:
Sounds good; thanks for the quick trial too!